Automattic/wp-calypso
# composite-checkout: Refactor Checkout steps as direct children

#### Changes proposed in this Pull Request

We need a way to run step validation when the "Continue" button is pressed (as opposed to on keypress) and to be able to run that validation asynchronously. #39087 was an attempt to do that, but because of issues with scope and state access, it kept running into problems. It also further complicates the API for steps.

This PR takes a different approach by refactoring the way steps are provided to the checkout form which should hopefully both solve the scope issues plaguing the previous attempt while at the same time making the step API easier to understand and use.

#### Background

Steps were conceived as "step objects" which were plain JS objects that followed a defined schema. Most of that schema was a collection of React elements to be displayed in different parts of the step, and in different situations. That part worked fine. However, in order to run validation on the step's data, the object also included a function called `isCompleteCallback`.

The original idea for that function was something simple that could be called on every render and declaratively return if the step was complete, but it quickly became challenging to in practice since the data in the step components was not easily accessible in the scope where the step was defined, and redefining the step could cause performance issues. Furthermore, the WPCOM contact form performs its validation with a server call which we cannot reasonably call on every render even if we had the ability to call `isCompleteCallback` asynchronously.

#39087 attempted to solve these issues by calling `isCompleteCallback` only when the "Continue" button was pressed, but the scope issues still remained.

#### A new step API

In this PR we remove the concept of step objects entirely, and instead rely on React components. The checkout form still begins with the `Checkout` component, but now that component has children. Any child is allowed but the `CheckoutStepBody` component can be used to render something that looks like a checkout step. Rather than having to write your own logic for these steps, you can instead use the `CheckoutSteps` component wrapping `CheckoutStep` components. 

Most of the properties of the old step objects now reside as props on `CheckoutStep`. In fact, existing step objects can remain and their properties passed as these props since the underlying API remains the same.

The notable difference with the new API is that `isCompleteCallback`, now a prop, will not be called on every render and will instead be called with the "Continue" button is pressed. If the function returns a Promise, then the "Continue" button will change to "Please wait…" until the Promise resolves. This is the same approach taken in #39087 with one major advantage: since the callback is now defined in JSX, it has direct access to the scope of its parent, which can provide all the form data that the callback needs in its closure.

#### Testing instructions

Even though it's only at the top level, this is a pretty major change, so we'll need to verify that the form still behaves as it did before. Most notably we need to make sure that the steps change correctly, that active/complete/incomplete steps all display properly, that editing a step works as expected, and that validation works as expected.

The biggest visible change of this PR is that the "Continue" button is no longer disabled when a step is incomplete. Pressing it when the step has invalid data should display an error in some way.

This PR doesn't yet use async validation in Calypso, but it makes it possible. Try running the demo (`npm run composite-checkout-demo`) to see it in action.

#### To do

- [x] Update demo to use new API
- [x] Update WPCheckout to use new API
- [x] Make sure WPContactForm displays validation errors
- [x] Refactor tests to use new API
- [x] Update README with new API
- [x] Remove unused components and hooks from old API
